### PR TITLE
Custom SecurityContexts for envoy DaemonSet and contour Deployment

### LINKED ISSUE #112: Configurable Security Context
**Please describe the problem you have**
Contour and the certgen job do not run on OpenShift using the default security context:
```
securityContext:
  runAsNonRoot: true
  runAsUser: 65534
  runAsGroup: 65534
```

Failure status condition of deploy/contour:
```
  - lastTransitionTime: "2020-11-13T21:44:15Z"
    lastUpdateTime: "2020-11-13T21:44:15Z"
    message: 'pods "contour-5475898957-" is forbidden: unable to validate against
      any security context constraint: [spec.containers[0].securityContext.runAsUser:
      Invalid value: 65534: must be in the ranges: [1000590000, 1000599999]]'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```

Either make `securityContext` configurable or introduce provider-specific extensions to `Contour` so this can be configured automatically.

/cc @jpeach @Miciah

### PR DESCRIPTION
**Pull request by dmorgan81**
Allow users to specify a security context for both envoy daemonsets and contour deployments. The default value is unprivileged and is equivalent to the following:

```
contourSecurityContext:
  runAsUser: 65534
  runAsGroup: 65534
  runAsNonRoot: true
```

Note that this does change the default behavior for envoy daemonsets. Currently an envoy daemonset has an empty security context, meaning it will run with the user/group specified in the envoy image, which is commonly root. Switching to a non-root user by default will prevent the envoy container from binding to a low port. Existing setups will need to be updated by adding a `envoySecurityContext` entry with the appropriate settings if they need to bind to a low port.

Fixes #112
Updates #378

### CODE REVIEW

**Comment by dmorgan81** [comment]
I've been trying to run the e2e tests locally but I'm getting inconsistent results, even on the main branch:

- TestContourNodePortService often hangs.
- TestGateway and TestGatewayOwnership sometimes fail. The contour pods associated with these tests are stuck in a crash loop with the only logging being `contour: error: invalid Contour configuration: invalid Gateway parameters specified: controllerName required, try --help`

I just added a change to the e2e tests to use an empty PodSecurityContext for the contour pods since port 80 appears to be in the mix.  However, I'm unable to get through the entire e2e test suite because of the above two issues.

**Comment by sunjayBhatia** [comment]
> I've been trying to run the e2e tests locally but I'm getting inconsistent results, even on the main branch:
> 
> * TestContourNodePortService often hangs.
> * TestGateway and TestGatewayOwnership sometimes fail. The contour pods associated with these tests are stuck in a crash loop with the only logging being `contour: error: invalid Contour configuration: invalid Gateway parameters specified: controllerName required, try --help`
> 
> I just added a change to the e2e tests to use an empty PodSecurityContext for the contour pods since port 80 appears to be in the mix. However, I'm unable to get through the entire e2e test suite because of the above two issues.

ah, this is probably not related to your change at all, i think this is related to this change in Contour being merged: https://github.com/projectcontour/contour/commit/9e98a7284e680a3f71c14d35195588175e7605a6

We need to fix this up on main and then you should be able to merge main and your change go green

File: internal/objects/deployment/deployment.go [hunk 1]
```diff
@@ -304,7 +304,6 @@ func DesiredDeployment(contour *operatorv1alpha1.Contour, image string) *appsv1.
 					ServiceAccountName:            objutil.ContourRbacName,
 					RestartPolicy:                 corev1.RestartPolicyAlways,
 					SchedulerName:                 "default-scheduler",
-					SecurityContext:               objutil.NewUnprivilegedPodSecurity(),
```
Comment (nak3) [inline internal/objects/deployment/deployment.go:307]: I think we don't need to move this line.
If `contour.ContourSecurityContextExists()` is `true`, we just overrides `deploy.Spec.Template.Spec.SecurityContext`.

File: api/v1alpha1/contour_types.go [hunk 2]
```diff
@@ -111,6 +111,16 @@ type ContourSpec struct {
 	//
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// ContourSecurityContext defines a PodSecurityContext for Contour pods.
+	//
+	// +optional
+	ContourSecurityContext *corev1.PodSecurityContext `json:"contourSecurityContext,omitempty"`
+
+	//EnvoySecurityContext defines a PodSecurityContext for Envoy pods.
```
Comment (nak3) [inline api/v1alpha1/contour_types.go:120]:
```suggestion
	// EnvoySecurityContext defines a PodSecurityContext for Envoy pods.
```

nit

File: api/v1alpha1/helpers.go [hunk 3]
```diff
@@ -95,3 +95,11 @@ func (c *Contour) EnvoyTolerationsExist() bool {
 
 	return false
 }
+
```
Comment (nak3) [inline api/v1alpha1/helpers.go:98]: I think `ContourSecurityContextExists` and `EnvoySecurityContextExists` should have a comment line as these are public functions. I wonder why lint check does not produce an error :thinking:

**Review by nak3 — COMMENTED** [review]
I also wanted this feature! Thank you :+1:

File: internal/objects/deployment/deployment.go (hunk 1 above)
Comment (dmorgan81) [reply internal/objects/deployment/deployment.go:307]: It might be premature optimization but I was hoping to avoid an unnecessary allocation.

**Comment by sunjayBhatia** [comment]
@dmorgan81 I think this PR might be placed on a lower priority to merge at the moment since the Operator/Contour relationship is going through a rework as part of our Gateway API support work. Apologies for that, but we will try to get to this soon!

**Review by sunjayBhatia — COMMENTED** [review]
Added https://github.com/projectcontour/contour/issues/4039 and https://github.com/projectcontour/contour/issues/4040 to track we want to make sure this is captured in the new world where Contour can manage the Envoy daemonset etc.

**Comment by arjunrn** [comment]
@dmorgan81 The jobs created by the operator will also have to use the custom security context. For [reference](https://github.com/openshift/contour-operator/pull/10).

**Comment by skriss** [comment]
This is stale, closing out.

**Pull request closed without merging**